### PR TITLE
extension: User space event

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
UserSpaceFileEventImpl provides 3 functions

1. setEnabled(events) declared by FileEvent. The events are registered for callback. Each call detects fired events once, and IO handle will notify the new events in the long run.
1. activate(events). Explicitly activate the events. The above enabling events are not honored.
1. poll(events). It should only be used by IO handle. Unlike activate, poll honors the setEnabled. Will schedule callback if there is no pending callback.

Additional Description:
Part of #13418
Risk Level: LOW all codes are in extension
Testing: Mocked userspace io handle to unit test userspace event.
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc
Comment (antoniovicente) [inline test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc:129]: Please also add:
```
{
  SCOPED_TRACE("no auto reschedule");
  EXPECT_CALL(ready_cb_, called(_)).Times(0);
  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
￼}
```

File: test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc
Comment (antoniovicente) [inline test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc:184]: Would be good to verify that activate works even in cases where the mask passed in to setEnabled does not contain the event being activated.

File: test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc
Comment (antoniovicente) [inline test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc:368]: I think you mean "but" instead of "by"

**Review by antoniovicente — COMMENTED** [review]
Looks pretty good.  Thanks so much for splitting this off and sending it out for review before the full change.

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc:42]: Hmm... I have no idea how the activate(Closed) is used. I can remove the `Closed` from the ASSERT for now and add the comment. I can run a full test later with the removal of the Closed here and the other FileEventImpl.

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.h
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.h:20]: Yeah... Inner class makes much more sense. Updating...

File: test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc
Comment (lambdai) [reply test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc:129]: :+1:

File: test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc
Comment (lambdai) [reply test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc:368]: Yes... Fixed

File: test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc
Comment (lambdai) [reply test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc:184]: That's a critical case. It may be hidden in some test cases. Let me create a new named test case

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.h
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.h:59]: Sorry for the unclear! The name `poll` is inspired by here. https://elixir.bootlin.com/linux/latest/source/include/linux/fs.h#L1832

Yes, activeateIfEnabled is a better name in our context. Updating...

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc:42]: Lots of tests are broken due to this change

**Review by lambdai — COMMENTED** [review]
1 round update. Namespace name dedup and inner class comes later

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc:78]: > Event::FileReadyType::Closed is never passed in to this method.

setEnabled should check if `Closed` has occurred.

Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc:78]: > setEnabled should check if `Closed` has occurred.

I was wrong about ^^.

setEnabled() could use this. But it doesn't and avoid checking enabled events twice.

Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc:78]: Anyway, setNewDataAvailable() should activate `Closed` to emulate shutdown(SHUT_WR) at peer so activateIfEnabled() must accept `Closed`. WDYT?

Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc:78]: Also, since activateIfEnabled() accept `Closed`, the activate() invoked by `activateIfEnabled()` should also accept `Closed`. Although `Network::Connection` may not directly call `activate(Closed)` as you described

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc]: Won't change:see activateIfEnabled()

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc]: renamed and moved all trivial method to header file

File: source/extensions/io_socket/user_space_io_socket/BUILD
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/BUILD]: add an empty config.h
the status is wip.
secruity_posture is "unknown".

What is the best secruity_posture indicating "don't use"?

**Comment by lambdai** [comment]
@antoniovicente  ready to be reviewed now

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (antoniovicente) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc]: The delivery of Closed events doesn't work on real sockets in some platforms.  I think use of Closed should be removed across the board eventually, but I guess for now it seems sensible to accept Closed for activateIfEnabled, but not for activate.  Seems like #13418 handles shutdown correctly by doing activateIfEnabled(Read | Closed); Closed is not always registered, sometimes you need to rely on the read registration to be the one that carries the event across.

File: source/extensions/io_socket/user_space/config.h
Comment (antoniovicente) [inline source/extensions/io_socket/user_space/config.h:7]: Please make include full link in comment.  Also, I think this is the wrong PR number.

```
// Will add config and factory in https://github.com/envoyproxy/envoy/pull/13865
```

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (antoniovicente) [inline source/extensions/io_socket/user_space/file_event_impl.cc:66]: nit: // Filter out disabled events.

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (antoniovicente) [inline source/extensions/io_socket/user_space/file_event_impl.cc:67]: nit: change filter_enabled to filtered_events

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (antoniovicente) [inline source/extensions/io_socket/user_space/file_event_impl.cc:43]: if ((events & Event::FileReadyType::Read) && (io_source_.isReadable() || io_source_.isPeerShutDownWrite()) {

File: source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space_io_socket/user_space_file_event_impl.cc]: > The delivery of Closed events doesn't work on real sockets in some platforms.
That's why the initial commits do not handle `Closed` :)

Ideally, the event owner (e.g. connection) can ask for `Closed` and consider `Closed` event as optimization. Especially when the L7 protocol doesn't have to drain the io_handle buffer. We can put this long term goal aside.

Even the connection goes as you described, this extra `Closed` delivery is not breaking the connection, since there are some platforms (mostly importantly, linux) could deliver `Closed` as this userspace file event :)

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space/file_event_impl.cc:43]: Aha, great catch! Fixing...

File: source/extensions/io_socket/user_space/config.h
Comment (lambdai) [reply source/extensions/io_socket/user_space/config.h]: Updated comment with the correct PR

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space/file_event_impl.cc]: Actually, in the interface I add the comment that peerShutdownWrite() imply `isReadable()`. But it's nice to have explicit calls.

```
  /**
   * @return true if the pending receive buffer is not empty or read_end is set.
   */
  virtual bool isReadable() const PURE;
```

File: test/per_file_coverage.sh
Comment (antoniovicente) [inline test/per_file_coverage.sh:53]: What death tests?  I see no death tests in file_event_impl_test.cc

Please add some tests that call registerEventIfEmulatedEdge and unregisterEventIfEmulatedEdge to get to 100% coverage.

https://storage.googleapis.com/envoy-pr/e4c7e87/coverage/source/extensions/io_socket/user_space/file_event_impl.h.gcov.html

Comment (lambdai) [reply test/per_file_coverage.sh:53]: registerEventIfEmulatedEdge() and unregisterEventIfEmulatedEdge() are supposed to be death test. Only IoSocketHandleImpl calls them and UserSpace::IoHandle won't call in #13418

Comment (lambdai) [reply test/per_file_coverage.sh:53]: I add death tests and NOT_IMPLEMENTED_GCOVR_EXCL_LINE to clarify it.

Comment (lambdai) [reply test/per_file_coverage.sh:53]: Coverage 

65/67 ~= 97.0 %

The missing two items are the 
registerEventIfEmulatedEdge and unregisterEventIfEmulatedEdge

Death tests do not help

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (antoniovicente) [inline source/extensions/io_socket/user_space/file_event_impl.cc:44]: expn EOS

**Review by antoniovicente — APPROVED** [review]
Thanks!

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space/file_event_impl.cc]: fixed

**Comment by lambdai** [comment]
The latest commit dismissed the approval though. That commit is merely a trivial comment update.

**Review by antoniovicente — APPROVED** [review]

**Comment by antoniovicente** [comment]
Matt is on paternity leave.  @envoyproxy/senior-maintainers could one of you take a look?

File: source/extensions/io_socket/user_space/file_event_impl.h
Comment (ggreenway) [inline source/extensions/io_socket/user_space/file_event_impl.h:9]: Is this used? I didn't see anything clearly using it.

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (ggreenway) [inline source/extensions/io_socket/user_space/file_event_impl.cc:38]: Why clear the events here? Is it the same reason as in `FileEventImpl::setEnabled()`? Please add a comment on why this is correct.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall. I'm excited to see this feature moving forward!

File: source/extensions/io_socket/user_space/file_event_impl.cc
Comment (lambdai) [reply source/extensions/io_socket/user_space/file_event_impl.cc:38]: Yes, the event owner is likely to see consistent behavior among various event implementation. Updating the comment

File: source/extensions/io_socket/user_space/file_event_impl.h
Comment (lambdai) [reply source/extensions/io_socket/user_space/file_event_impl.h:8]: :+1:

File: source/extensions/io_socket/user_space/file_event_impl.h
Comment (lambdai) [reply source/extensions/io_socket/user_space/file_event_impl.h:9]: Yeah... event_impl_base is tied to libevent raw event. Stripping

**Comment by lambdai** [comment]
@ggreenway  Thanks! I address the included headers and also cleaned the bazel deps.

**Review by ggreenway — APPROVED** [review]
LGTM

**Comment by lambdai** [comment]
@ggreenway I believe the failing macos test is unrelated. It's failing consistently on other merged PRs. 
What is the missing part to merge this PR? Thanks!

**Comment by ggreenway** [comment]
Mac CI fix was merged an hour ago. Please merge main and try again, and it should work.

**Comment by lambdai** [comment]
@ggreenway All green now!

**Comment by moderation** [comment]
Interested in what this functionally is for. Userspace QUIC for HTTP/3? For use with gVisor's userspace Netstack? Something else?

**Comment by antoniovicente** [comment]
See https://github.com/envoyproxy/envoy/issues/11725

This is a step towards generalized connect handling; one use case could be access to internal listeners which are reachable via HTTP CONNECT, and support of the CONNECT method over HTTP2.

**Pull request merged**
